File: summary/ts/22_annealing.ts
function simulatedAnnealing(initialSolution: number[], costFunction: (arr: number[]) => number): number[] {
    let currentSolution: number[] = initialSolution.slice();
    let currentCost: number = costFunction(currentSolution);
    let temperature: number = 100.0;
    const minTemperature: number = 0.01;
    const alpha: number = 0.9;

    while (temperature > minTemperature) {
        let nextSolution: number[] = currentSolution.slice();
        const i: number = Math.floor(Math.random() * currentSolution.length);
        const j: number = Math.floor(Math.random() * currentSolution.length);
        [nextSolution[i], nextSolution[j]] = [nextSolution[j], nextSolution[i]];
        const nextCost: number = costFunction(nextSolution);

        if (nextCost < currentCost || Math.exp((currentCost - nextCost) / temperature) > Math.random()) {
            currentSolution = nextSolution;
            currentCost = nextCost;
        }

        temperature *= alpha;
    }

    return currentSolution;
}

// Example usage:
const inputArray: number[] = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5];
const optimizedArray: number[] = simulatedAnnealing(inputArray, (arr) => arr.reduce((sum, num) => sum + num ** 2, 0));
console.log(`Optimized Array using Simulated Annealing: ${optimizedArray}`);

File: summary/dart/22_annealing.dart
import 'dart:math';

List<int> simulatedAnnealing(List<int> initialSolution, int Function(List<int>) costFunction) {
  List<int> currentSolution = List.from(initialSolution);
  int currentCost = costFunction(currentSolution);
  double temperature = 100.0;
  double minTemperature = 0.01;
  double alpha = 0.9;
  Random random = Random();

  while (temperature > minTemperature) {
    List<int> nextSolution = List.from(currentSolution);
    int i = random.nextInt(currentSolution.length);
    int j = random.nextInt(currentSolution.length);
    int temp = nextSolution[i];
    nextSolution[i] = nextSolution[j];
    nextSolution[j] = temp;
    int nextCost = costFunction(nextSolution);

    if (nextCost < currentCost || exp((currentCost - nextCost) / temperature) > random.nextDouble()) {
      currentSolution = List.from(nextSolution);
      currentCost = nextCost;
    }

    temperature *= alpha;
  }

  return currentSolution;
}

void main() {
  // Example usage:
  List<int> inputArray = [5, 3, 8, 6, 2];
  List<int> optimizedArray = simulatedAnnealing(inputArray, (arr) => arr.fold(0, (sum, num) => sum + num * num));
  print("Optimized Array using Simulated Annealing: $optimizedArray");
}
